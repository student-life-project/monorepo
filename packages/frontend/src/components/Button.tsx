import { memo } from 'react';
import xw from 'xwind';
import styled from '@emotion/styled';

type IButton = {
  FPrimary?: boolean;
  FSecondary?: boolean;
  FSuccess?: boolean;
  FDanger?: boolean;
  FWarning?: boolean;
  BPrimary?: boolean;
  BSecondary?: boolean;
  BSuccess?: boolean;
  BDanger?: boolean;
  BWarning?: boolean;
  disabled?: boolean;
  round?: boolean;
  large?: boolean;
  small?: boolean;
  children: React.ReactNode;
} & React.ButtonHTMLAttributes<HTMLButtonElement>;

const Btn = styled.button<IButton>`
  ${xw`
    flex
    px-5 
    py-2.5
    text-sm 
    font-bold
    text-white 
    rounded-md
    focus:ring
    items-center
    duration-500
    transition ease-in
    focus:outline-none
    focus:ring-offset-1
    focus:border-blue-300
  `}

  ${({ FPrimary }) => FPrimary && xw`bg-blue-500 hover:bg-blue-600`}

  ${({ FSecondary }) => FSecondary && xw`bg-gray-500 hover:bg-gray-600`}

  ${({ FSuccess }) => FSuccess && xw`bg-green-500 hover:bg-green-600`}

  ${({ FDanger }) => FDanger && xw`bg-red-500 hover:bg-red-600`}

  ${({ FWarning }) => FWarning && xw`bg-yellow-500 hover:bg-yellow-600`}

  ${({ BPrimary }) =>
    BPrimary && xw`text-blue-600 border border-blue-600 hover:bg-blue-100`}

  ${({ BSecondary }) =>
    BSecondary && xw`text-gray-600 border border-gray-600 hover:bg-gray-100`}

  ${({ BSuccess }) =>
    BSuccess && xw`text-green-600 border border-green-600 hover:bg-green-100`}

  ${({ BDanger }) =>
    BDanger && xw`text-red-600 border border-red-600 hover:bg-red-100`}
    
  ${({ BWarning }) =>
    BWarning &&
    xw`text-yellow-600 border border-yellow-600 hover:bg-yellow-100`}

  ${({ disabled }) =>
    disabled &&
    xw`text-gray-600 border border-gray-300 bg-gray-300 hover:bg-gray-300 cursor-not-allowed`}

  ${({ round }) => round && xw`rounded-full`}

  ${({ large }) => large && xw`text-lg py-3 px-6`}

  ${({ small }) => small && xw`text-xs py-2 px-4`}
`;

const Button: React.FC<IButton> = ({ children, ...props }) => (
  <Btn {...props}>{children}</Btn>
);

export default memo(Button);
