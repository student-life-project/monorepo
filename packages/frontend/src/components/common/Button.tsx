// eslint-disable-next-line simple-import-sort/imports
import xw from 'xwind';
import styled from '@emotion/styled';
import { FC } from 'react';

type TButton = {
  FPrimary?: boolean;
  FSecondary?: boolean;
  FSuccess?: boolean;
  FDanger?: boolean;
  FWarning?: boolean;
  BPrimary?: boolean;
  BSecondary?: boolean;
  BSuccess?: boolean;
  BDanger?: boolean;
  BWarning?: boolean;
  disabled?: boolean;
  round?: boolean;
  large?: boolean;
  small?: boolean;
  like?: boolean;
  dislike?: boolean;
  children: React.ReactNode;
} & React.ButtonHTMLAttributes<HTMLButtonElement>;

const Btn = styled.button<TButton>`
  ${xw`
    flex
    px-5
    py-2.5
    text-sm
    font-bold
    text-white
    rounded-md
    focus:ring
    items-center
    duration-500
    items-center
    justify-center
    transition ease-in
    focus:outline-none
    focus:ring-offset-1
    focus:border-blue-300
  `}

  ${({ FPrimary }) => FPrimary && xw`bg-primary hover:opacity-75`}

  ${({ FSecondary }) => FSecondary && xw`bg-secondary-1 hover:bg-gray-600`}

  ${({ FSuccess }) => FSuccess && xw`bg-green-500 hover:bg-green-600`}

  ${({ FDanger }) => FDanger && xw`bg-red-500 hover:bg-red-600`}

  ${({ FWarning }) => FWarning && xw`bg-yellow-500 hover:bg-yellow-600`}

  ${({ BPrimary }) =>
    BPrimary && xw`text-primary border border-primary hover:bg-blue-100`}

  ${({ BSecondary }) =>
    BSecondary &&
    xw`text-secondary-1 border border-secondary-1 hover:bg-gray-100`}

  ${({ BSuccess }) =>
    BSuccess && xw`text-green-600 border border-green-600 hover:bg-green-100`}

  ${({ BDanger }) =>
    BDanger && xw`text-red-600 border border-red-600 hover:bg-red-100`}
    
  ${({ BWarning }) =>
    BWarning &&
    xw`text-yellow-600 border border-yellow-600 hover:bg-yellow-100`}

  ${({ disabled }) =>
    disabled &&
    xw`text-gray-600 border border-gray-300 bg-gray-300 hover:bg-gray-300 cursor-not-allowed`}

  ${({ round }) => round && xw`rounded-full`}

  ${({ large }) => large && xw`text-lg py-3 px-6`}

  ${({ small }) => small && xw`text-xs py-2 px-4`}

  ${({ like }) => like && xw`bg-blue-100`}
`;

const Button: FC<TButton> = ({ children, ...props }) => (
  <Btn {...props}>{children}</Btn>
);

export default Button;
