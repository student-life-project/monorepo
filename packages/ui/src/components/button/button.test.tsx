import { fireEvent, render } from '@testing-library/react';

import Button, { Props } from '.';

describe('Button', () => {
  let props: Props;

  beforeEach(() => {
    props = {
      children: <div>I am a button</div>,
      onClick: jest.fn(),
      disabled: false,
      type: 'submit',
      color: '',
      dataTestId: 'button',
    };
  });

  describe('actions', () => {
    it('triggers the callback when clicked', () => {
      const { getByTestId } = render(<Button {...props} />);
      const button = getByTestId('button');

      fireEvent.click(button);

      expect(props.onClick).toHaveBeenCalledTimes(1);
    });

    it('does not trigger the callback when clicked if the button is disabled', () => {
      props.disabled = true;
      const { getByTestId } = render(<Button {...props} />);
      const button = getByTestId('button');

      fireEvent.click(button);

      expect(props.onClick).toHaveBeenCalledTimes(0);
    });
  });

  describe('render()', () => {
    it('renders a submit button', () => {
      const { container } = render(<Button {...props} />);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('renders a reset button', () => {
      props.type = 'reset';

      const { container } = render(<Button {...props} />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
