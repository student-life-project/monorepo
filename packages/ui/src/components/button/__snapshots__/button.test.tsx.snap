// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button render() renders a reset button 1`] = `
<button
  class="css-17a9ggc"
  color=""
  data-testid="button"
  type="reset"
>
  <span>
    <div>
      I am a button
    </div>
  </span>
</button>
`;

exports[`Button render() renders a submit button 1`] = `
<button
  class="css-17a9ggc"
  color=""
  data-testid="button"
  type="submit"
>
  <span>
    <div>
      I am a button
    </div>
  </span>
</button>
`;
